### PR TITLE
switch to using pointers to structs in fields

### PR DESCRIPTION
**Pull request by JackQuincy**
This is so that we can omit fields in output in the managed azure service.

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/455)

### CODE REVIEW

**Comment by JackQuincy** [comment]
@acs-bot test this please

File: pkg/api/vlabs/validate.go [hunk 1]
```diff
@@ -25,7 +25,7 @@ func (o *OrchestratorProfile) Validate() error {
 	}
 
 	if o.OrchestratorType != Kubernetes &&
-		(o.KubernetesConfig.KubernetesImageBase != "" || o.KubernetesConfig.NetworkPolicy != "") {
+		(o.KubernetesConfig != nil) {
```
Comment (amanohar) [inline pkg/api/vlabs/validate.go]: The case of KubernetesConfig being present and KubernetesImageBase/NetworkPolicy  should also be handled.

**Comment by weinong** [comment]
Reviewed 6 of 13 files at r1.
Review status: 6 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/api/convertertoapi.go, line 391 at r1](https://reviewable.io:443/reviews/azure/acs-engine/455#-KgWJNjSHkcy_1BvEQmX:-KgWJNjSHkcy_1BvEQmY:b-i17vjz) ([raw file](https://github.com/azure/acs-engine/blob/16421b9bb1cf73021eff33cf423ad6954d7531d9/pkg/api/convertertoapi.go#L391)):*
> ```Go
> func convertV20160330VMDiagnostics(v20160330 *v20160330.VMDiagnostics, api *VMDiagnostics) {
> 	api.Enabled = v20160330.Enabled
> 	api.StorageURL = v20160330.StorageURL
> ```

StorageURL needs to be a pointer, and copies when it's not nil.

---

*[pkg/api/types.go, line 182 at r1](https://reviewable.io:443/reviews/azure/acs-engine/455#-KgWIslv-PasjmOcHQKB:-KgWIslv-PasjmOcHQKC:b7bb0ew) ([raw file](https://github.com/azure/acs-engine/blob/16421b9bb1cf73021eff33cf423ad6954d7531d9/pkg/api/types.go#L182)):*
> ```Go
> 	// This field is readonly as ACS RP will create a storage account
> 	// for the customer.
> 	StorageURL neturl.URL `json:"storageUrl"`
> ```

I think it needs to be a pointer

---

*[pkg/api/v20160330/types.go, line 146 at r1](https://reviewable.io:443/reviews/azure/acs-engine/455#-KgWIn6CLF3UtC0TA2a-:-KgWIn6CLF3UtC0TA2a0:b7bb0ew) ([raw file](https://github.com/azure/acs-engine/blob/16421b9bb1cf73021eff33cf423ad6954d7531d9/pkg/api/v20160330/types.go#L146)):*
> ```Go
> 	// This field is readonly as ACS RP will create a storage account
> 	// for the customer.
> 	StorageURL neturl.URL `json:"storageUrl"`
> ```

I think it needs to be a pointer

---

*[pkg/api/v20170131/types.go, line 159 at r1](https://reviewable.io:443/reviews/azure/acs-engine/455#-KgWIRplyRaVKMmAOyuY:-KgWIRpmnTZQz9zhba5t:bdn56xi) ([raw file](https://github.com/azure/acs-engine/blob/16421b9bb1cf73021eff33cf423ad6954d7531d9/pkg/api/v20170131/types.go#L159)):*
> ```Go
> 	// This field is readonly as ACS RP will create a storage account
> 	// for the customer.
> 	StorageURL neturl.URL `json:"storageUrl"`
> ```

I think it needs to be a pointer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/acs-engine/455)*

File: pkg/api/vlabs/validate.go [hunk 2]
```diff
@@ -263,6 +263,9 @@ func (a *Properties) validateNetworkPolicy() error {
 
 	switch a.OrchestratorProfile.OrchestratorType {
 	case Kubernetes:
+		if a.OrchestratorProfile.KubernetesConfig == nil {
+			return nil
+		}
```
Comment (amanohar) [inline pkg/api/vlabs/validate.go]: Should it return nil or empty since Go's default string value is empty?

File: pkg/api/vlabs/validate.go (hunk 1 above)
Comment (JackQuincy) [inline pkg/api/vlabs/validate.go]: good catch

File: pkg/api/vlabs/validate.go (hunk 2 above)
Comment (JackQuincy) [inline pkg/api/vlabs/validate.go]: this is returning an error as this is validation. so returning nil means this is a valid value. I'll skip the rest of this case which will leave networkPolicy as "" which is how this method behaves today.

**Comment by JackQuincy** [comment]
Good idea on the more simple test, that doesn't depend on our test inputs as much. I ran classic/kubernetes.classic.json through and didn't see unexpected objects.

File: pkg/api/convertertoapi.go [hunk 3]
```diff
@@ -307,7 +382,8 @@ func convertV20160930VMDiagnostics(v20160930 *v20160930.VMDiagnostics, api *VMDi
 }
 
 func convertV20160330DiagnosticsProfile(v20160330 *v20160330.DiagnosticsProfile, api *DiagnosticsProfile) {
-	convertV20160330VMDiagnostics(&v20160330.VMDiagnostics, &api.VMDiagnostics)
+	api.VMDiagnostics = &VMDiagnostics{}
+	convertV20160330VMDiagnostics(v20160330.VMDiagnostics, api.VMDiagnostics)
```
Comment (weinong) [inline pkg/api/convertertoapi.go]: nil check on v20160330.VMDiagnostics?

Comment (JackQuincy) [inline pkg/api/convertertoapi.go]: It doesn't makes sense to not have it if you have a diagnostics profile, it is the only field. So I turned off omit empty. But I guess we could be called without it.

Comment (weinong) [inline pkg/api/convertertoapi.go]: stupid GitHub doesn't allow me to leave comments elsewhere. I meant nil check on v20160330.VMDiagnostics inside convertV20160330VMDiagnostics

**Comment by JackQuincy** [comment]
@acs-bot test this please

**Pull request merged**